text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,970 @@
+
+## `ops::ControlFlow`
+
+```rust
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ControlFlow<B, C = ()> {
+    /// Exit the operation without running subsequent phases.
+    Break(B),
+    /// Move on to the next phase of the operation as normal.
+    Continue(C),
+}
+```
+
+## The traits
+
+```rust
+pub trait Try: FromTryResidual {
+    /// The type of the value consumed or produced when not short-circuiting.
+    type Output;
+
+    /// A type that "colours" the short-circuit value so it can stay associated
+    /// with the type constructor from which it came.
+    type Residual;
+
+    /// Used in `try{}` blocks to wrap the result of the block.
+    fn from_output(x: Self::Output) -> Self;
+
+    /// Determine whether to short-circuit (by returning `ControlFlow::Break`)
+    /// or continue executing (by returning `ControlFlow::Continue`).
+    fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
+}
+
+pub trait FromTryResidual<Residual = <Self as Try>::Residual> {
+    /// Recreate the `Try` type from a related residual
+    fn from_residual(x: Residual) -> Self;
+}
+```
+
+## Desugaring `?`
+
+The previous desugaring of `x?` was
+
+```rust
+match Try::into_result(x) {
+	Ok(v) => v,
+	Err(e) => return Try::from_error(From::from(e)),
+}
+```
+
+The new one is very similar:
+
+```rust
+match Try::branch(x) {
+	ControlFlow::Continue(v) => v,
+	ControlFlow::Break(r) => return FromTryResidual::from_residual(r),
+}
+```
+
+It's just left conversion (such as `From::from`) up to the implementation instead of forcing it in the desugar.
+
+## Standard implementations
+
+### `Result`
+
+```rust
+impl<T, E> ops::Try for Result<T, E> {
+    type Output = T;
+    type Residual = Result<!, E>;
+
+    #[inline]
+    fn from_output(c: T) -> Self {
+        Ok(c)
+    }
+
+    #[inline]
+    fn branch(self) -> ControlFlow<Self::Residual, T> {
+        match self {
+            Ok(c) => ControlFlow::Continue(c),
+            Err(e) => ControlFlow::Break(Err(e)),
+        }
+    }
+}
+
+impl<T, E, F: From<E>> ops::FromTryResidual<Result<!, E>> for Result<T, F> {
+    fn from_residual(x: Result<!, E>) -> Self {
+        match x {
+            Err(e) => Err(From::from(e)),
+        }
+    }
+}
+```
+
+### `Option`
+
+```rust
+impl<T> ops::Try for Option<T> {
+    type Output = T;
+    type Residual = Option<!>;
+
+    #[inline]
+    fn from_output(c: T) -> Self {
+        Some(c)
+    }
+
+    #[inline]
+    fn branch(self) -> ControlFlow<Self::Residual, T> {
+        match self {
+            Some(c) => ControlFlow::Continue(c),
+            None => ControlFlow::Break(None),
+        }
+    }
+}
+
+impl<T> ops::FromTryResidual for Option<T> {
+    fn from_residual(x: <Self as ops::Try>::Residual) -> Self {
+        match x {
+            None => None,
+        }
+    }
+}
+```
+
+### `Poll`
+
+These reuse `Result`'s residual type, so don't need additional `FromTryResidual` implementations on `Result`.
```

Comment:
```suggestion
These reuse `Result`'s residual type, so we don't need additional `FromTryResidual` implementations on `Result` to allow interconversion.
```

Reply:
Actually, should these impls be generalised to all types that implement `Try`?

```rust
impl<T: Try> Try for Poll<T> { ... }
impl<T: TryFromResidual<R>, R> FromTryResidual<R> for Poll<T> { ... }
```

Reply:
(FWIW this is exactly an example of why I dislike the `FromTry` name, since you seem to have accidentally written `TryFrom` here out of habit.)

Reply:
Which is another reason I'd like the `bubble` name for the residual/error/break type.

The second impl would then be:

```rust
impl<T: FromBubble<R>, R> FromBubble<R> for Poll<T> { ... }
```